// @astrojs/vercel/dist/image/dev-service.js
import squooshService from "astro/assets/services/squoosh";
import { sharedValidateOptions } from "./shared";
const service = {
  validateOptions: (options, serviceOptions) => sharedValidateOptions(options, serviceOptions, "development"),
  getHTMLAttributes(options, serviceOptions) {
    const { inputtedWidth, ...props } = options;
    if (inputtedWidth) {
      props.width = inputtedWidth;
    }
    return squooshService.getHTMLAttributes(props, serviceOptions);
  },
  getURL(options) {
    const fileSrc = typeof options.src === "string" ? options.src : options.src.src;
    const searchParams = new URLSearchParams();
    searchParams.append("href", fileSrc);
    options.width && searchParams.append("w", options.width.toString());
    options.quality && searchParams.append("q", options.quality.toString());
    return "/_image?" + searchParams;
  },
  parseURL(url) {
    const params = url.searchParams;
    if (!params.has("href")) {
      return void 0;
    }
    const transform = {
      src: params.get("href"),
      width: params.has("w") ? parseInt(params.get("w")) : void 0,
      quality: params.get("q")
    };
    return transform;
  },
  transform(inputBuffer, transform, serviceOptions) {
    transform.format = "webp";
    return squooshService.transform(inputBuffer, transform, serviceOptions);
  }
};
var dev_service_default = service;
export {
  dev_service_default as default
};
